### PR TITLE
✨ Package and publish rancher turtles on each release

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What this PR does / why we need it**:

- introduces release workflow to run as a GH action similar to what is done both in CAPI/CAPI Operator which will be triggered on a new release tag push. This workflow will create a draft release and publish a helm chart. We will host our repository using github pages and publish helm charts using helm/chart-releaser-action action.
- reworks kustomize folders to adjust it to charts introduction and created a new set of kustomize configurations
- introduces new Makefile targets for release / releasing charts 

This repository will also be used as a helm chart repository, we will host index.yaml in root of project and update it with every release. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes: https://github.com/rancher-sandbox/rancher-turtles/issues/14

**Special notes for your reviewer**:

**Checklist**:

- [X] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
> Generally looks good to me.
> 
> With the changes to the `config` folder have you tried running `make dev-env` to check that it still works?

@richardcase thanks for review and pointing out. I was able to run it properly including my last commit (https://github.com/rancher-sandbox/rancher-turtles/pull/34/commits/077b53d21be5dd3ed3c399e0b794aec3f8bc18fd) to dev script (had to use newer version of rancher 2.7.5 instead of 2.7.3 since kind uses kindest/node:v1.26.3 image). 
The same commit includes a new makefile target to cleanup development environment. 

Screenshot 2023-06-20 at 21 26 19

**Comment by furkatgofurov7** [comment]
> Generally looks good to me.
> 
> With the changes to the `config` folder have you tried running `make dev-env` to check that it still works?

@richardcase thanks for review and pointing out. I was able to run it properly including my last commit (https://github.com/rancher-sandbox/rancher-turtles/pull/34/commits/41f5d2a9ac9b8e4b5427df52ecaeeffeb098e259) to dev script (had to use newer version of rancher 2.7.5 instead of 2.7.3 since kind uses kindest/node:v1.26.3 image). 
The same commit includes a new makefile target to cleanup development environment. 


Screenshot 2023-07-13 at 17 00 43

**Review by Danil-Grigorev — COMMENTED** [review]
Looks good, only some questions

File: hack/chart-update/main.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+)
+
+const (
+	indexFilePath = "../../index.yaml"
+	gitHubOrgName = "rancher-sandbox"
+	repoName      = "rancher-turtles"
+)
+
+func main() {
+	fmt.Println("🚀 Starting index.yaml update tool")
```
[the remark this answers is not in the document]
Comment (alexander-demicev) [reply hack/chart-update/main.go:28]: emojis are fun

File: hack/chart-update/main.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		fmt.Println("❌ Error loading chart: ", err)
+		os.Exit(1)
+	}
+
+	return archivePath, chart
+}
+
+func addEntryToIndexFile(indexFile *repo.IndexFile, chartAsset *github.ReleaseAsset, archivePath string, chart *chart.Chart) {
+	s := strings.Split(*chartAsset.BrowserDownloadURL, "/") // https://github.com/helm/chart-releaser/blob/main/pkg/releaser/releaser.go#L299
```
[the remark this answers is not in the document]
Comment (alexander-demicev) [reply hack/chart-update/main.go:144]: We need a link to the github release like here https://github.com/kubernetes-sigs/cluster-api-operator/blob/main/index.yaml#L12, this is how helm will know where to look for package

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      - name: Set env
+        run:  echo "RELEASE_TAG=${GITHUB_REF:10}" >> $GITHUB_ENV
+      - name: checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: '1.19'
```
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply .github/workflows/release.yaml]: Done, bumped to 1.20/v1.20.4

File: hack/chart-update/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply hack/chart-update/main.go:28]: :)

**Comment by furkatgofurov7** [comment]
> > Generally looks good to me.
> > With the changes to the `config` folder have you tried running `make dev-env` to check that it still works?
> 
> @richardcase thanks for review and pointing out. I was able to run it properly, including my last commit ([41f5d2a](https://github.com/rancher-sandbox/rancher-turtles/commit/41f5d2a9ac9b8e4b5427df52ecaeeffeb098e259)) which slightly changes dev script (had to use newer version of rancher 2.7.5 instead of 2.7.3 since kind uses kindest/node:v1.26.3 image). The same commit includes a new makefile target to cleanup development environment.
> 
> Screenshot 2023-07-13 at 17 00 43

Aplogies, did not pay attention to failure messages 👀 , I'll investigate if kustomize changes is the cause of it

File: config/chart/manager_image_patch.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: controller-manager
+  namespace: system
+spec:
+  template:
+    spec:
+      containers:
+        - image: ghcr.io/rancher-turtles:
```
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply config/chart/manager_image_patch.yaml]: Fixed

**Comment by furkatgofurov7** [comment]
@richardcase @Danil-Grigorev as discussed offline, this should be ready for re-review again, rancher turtles controller pod not becoming ready does not seem to be related to these changes based on my testing (suspected commit is https://github.com/rancher-sandbox/rancher-turtles/pull/29/commits/9d132ec7e41f5fb645d5be96d7e1144013896c6e)

**Review by alexander-demicev — APPROVED** [review]

**Comment by alexander-demicev** [comment]
thanks for doing this!

**Review by Danil-Grigorev — APPROVED** [review]

**Pull request merged**